internal/iso_packaging/iso_packaging.go

```diff
@@ -0,0 +1,241 @@
+	"strings"
+
+	"github.com/project-flotta/osbuild-operator/api/v1alpha1"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/aws-sdk-go/aws"
+)
+
+const (
+	templateCommand string = "isopackage %s -ks %s --upload-target"
+	kickstart       string = "init.ks"
+)
+
+var (
+	jobTTLafterFinish int32 = 100   // Job will be deleted after finished with this
+	deadlineSeconds   int64 = 36000 // Job will be terminated after a hour.
+)
+
+// Builder is a struct that manages a build to package an iso
+type Builder struct {
+	client      client.Client
+	jobSpec     *batchv1.Job
+	build       *v1alpha1.OSBuild
+	buildConfig *v1alpha1.OSBuildEnvConfig
+	baseImage   string
+}
+
+func NewBuilderJob(client client.Client, build *v1alpha1.OSBuild, buildConfig *v1alpha1.OSBuildEnvConfig, image string) (*Builder, error) {
+
+	if build == nil || buildConfig == nil {
+		return nil, fmt.Errorf("Invalid config added")
+	}
+
+	return &Builder{
+		client:      client,
+		jobSpec:     nil,
+		build:       build,
+		buildConfig: buildConfig,
+		baseImage:   image,
+	}, nil
+}
+
+func (b *Builder) Start() error {
+	sourceISO := b.build.Status.ComposerIso
+	if sourceISO == "" {
+		return fmt.Errorf("Cannot parse invalid iso image")
+	}
+	_, err := url.Parse(sourceISO)
+	if err != nil {
+		return err
+	}
+
+	if b.build.Spec.Kickstart == nil {
+		return fmt.Errorf("Kickstart is not defined, skipping")
+	}
+
+	command := fmt.Sprintf(templateCommand, b.build.Status.ComposerIso, kickstart)
+
+	jobSpec := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      b.build.Name,
+			Namespace: b.build.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: b.build.APIVersion,
+					Kind:       b.build.Kind,
+					Name:       b.build.Name,
+					UID:        b.build.UID,
+				},
+			},
+		},
+		Spec: batchv1.JobSpec{
+			TTLSecondsAfterFinished: &jobTTLafterFinish,
+			ActiveDeadlineSeconds:   &deadlineSeconds,
+			Template: v1.PodTemplateSpec{
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:    "iso-command",
+							Image:   b.baseImage,
+							Command: strings.Split(command, " "),
+							VolumeMounts: []v1.VolumeMount{
+								{
+									Name:      "config",
+									ReadOnly:  true,
+									MountPath: "/opt/iso_package/",
+								}},
+						},
+					},
+					RestartPolicy: v1.RestartPolicyNever,
+					Volumes: []v1.Volume{
+						{
+							Name: "config",
+							VolumeSource: v1.VolumeSource{
+								Projected: &v1.ProjectedVolumeSource{
+									Sources: []v1.VolumeProjection{{
+										ConfigMap: &v1.ConfigMapProjection{
+											LocalObjectReference: v1.LocalObjectReference{
+												Name: b.build.Spec.Kickstart.Name,
+											},
+										},
+									}},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if b.buildConfig.Spec.S3Service.AWS != nil {
```

Comment:
Since we force an upload, these `if` statements should be `if` `else` with a failure at the end

Reply:
No, if AWS is not in there does not mean that Generics3 can be in there, because is a pointer.

Reply:
I meant `if AWS` ... `else if S3` ... `else error`. I think that if we don't return an error on the third case, the Job will fail instead